common/src/main/java/io/druid/collections/DefaultBlockingPool.java

```diff
@@ -160,53 +159,39 @@ private T takeObject() throws InterruptedException
   }
 
   @Override
-  public ReferenceCountingResourceHolder<List<T>> takeBatch(final int elementNum, final long timeoutMs)
+  public List<ReferenceCountingResourceHolder<T>> takeBatch(final int elementNum, final long timeoutMs)
   {
     Preconditions.checkArgument(timeoutMs >= 0, "timeoutMs must be a non-negative value, but was [%s]", timeoutMs);
     checkInitialized();
     try {
-      return wrapObjects(timeoutMs > 0 ? pollObjects(elementNum, timeoutMs) : pollObjects(elementNum));
+      final List<T> objects = timeoutMs > 0 ? pollObjects(elementNum, timeoutMs) : pollObjects(elementNum);
+      return objects.stream().map(this::wrapObject).collect(Collectors.toList());
     }
     catch (InterruptedException e) {
       throw Throwables.propagate(e);
     }
   }
 
   @Override
-  public ReferenceCountingResourceHolder<List<T>> takeBatch(final int elementNum)
+  public List<ReferenceCountingResourceHolder<T>> takeBatch(final int elementNum)
   {
     checkInitialized();
     try {
-      return wrapObjects(takeObjects(elementNum));
+      return takeObjects(elementNum).stream().map(this::wrapObject).collect(Collectors.toList());
```

Comment:
I think this has a buffer leak (and it looks like the old code had the leak too, and so does `pollObjects`). If either `pollObjects` or `takeObjects` is interrupted while it's waiting for more objects to become available, then the objects popped from `objects` are not returned to the pool - they are lost.

Reply:
Would you elaborate more on how resource leak occurs on interruption?

The implementation of `takeBatch` is

```java
  private List<T> takeObjects(int elementNum) throws InterruptedException
  {
    final List<T> list = new ArrayList<>(elementNum);
    final ReentrantLock lock = this.lock;
    lock.lockInterruptibly();
    try {
      while (objects.size() < elementNum) {
        notEnough.await();
      }
      for (int i = 0; i < elementNum; i++) {
        list.add(objects.pop());
      }
      return list;
    }
    finally {
      lock.unlock();
    }
  }
```

and `InterruptedException` can be thrown at `lock.lockInterruptibly()` and `notEnough.await()`. `list.add(objects.pop())` is called only when there are enough number of available objects. 

`wrapObject()` also doesn't check the interruption state, so objects should be wrapped once `takeObjects()` returns them.

Reply:
Ah, you're right, as long as nothing involved checks interrupts: `list.add`, `objects.pop`, `wrapObject`, etc. It looks like that is the case so there is no leak. Nevermind.